libbeat/autodiscover/providers/kubernetes/node.go

```diff
@@ -0,0 +1,219 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package kubernetes
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/gofrs/uuid"
+	"k8s.io/api/core/v1"
+	k8s "k8s.io/client-go/kubernetes"
+
+	"github.com/elastic/beats/libbeat/autodiscover/builder"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/common/bus"
+	"github.com/elastic/beats/libbeat/common/kubernetes"
+	"github.com/elastic/beats/libbeat/common/safemapstr"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+type node struct {
+	uuid    uuid.UUID
+	config  *Config
+	metagen kubernetes.MetaGenerator
+	logger  *logp.Logger
+	publish func(bus.Event)
+	watcher kubernetes.Watcher
+}
+
+// NewNodeEventer creates an eventer that can discover and process node objects
+func NewNodeEventer(uuid uuid.UUID, cfg *common.Config, client k8s.Interface, publish func(event bus.Event)) (Eventer, error) {
+	metagen, err := kubernetes.NewMetaGenerator(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	logger := logp.NewLogger("autodiscover.node")
+
+	config := defaultConfig()
+	err = cfg.Unpack(&config)
+	if err != nil {
+		return nil, err
+	}
+
+	// Ensure that host is set correctly whenever the scope is set to "host". Make sure that host is empty
+	// when cluster scope is enforced.
+	if config.Scope == "host" {
+		config.Host = kubernetes.DiscoverKubernetesNode(config.Host, kubernetes.IsInCluster(config.KubeConfig), client)
+	} else {
+		config.Host = ""
+	}
+
+	logger.Debugf("Initializing a new Kubernetes watcher using host: %v", config.Host)
+
+	watcher, err := kubernetes.NewWatcher(client, &kubernetes.Node{}, kubernetes.WatchOptions{
+		SyncTimeout: config.SyncPeriod,
+		Node:        config.Host,
```

Comment:
I think all these should use config.Node, right? same happens in other eventers. Maybe we can rename `Host` to `HostDeprecated` to avoid its usage?